File: windows/internal/pairing/pairing.go
// Package pairing follows a LAN join request (docs/local-first.md §4): a
// new device names itself with POST /api/device, gets a four-digit code and
// waits, polling GET /api/me, until the owner allows it from one of their
// devices (the "pending" flag goes) or denies it (the device is gone).
package pairing

import (
	"context"
	"errors"
	"time"

	"github.com/Ferinmtk/droplet/windows/internal/hub"
)

// State is where a join request stands.
type State string

const (
	// Pending: waiting for the owner to allow it.
	Pending State = "pending"
	// Approved: let in; the token now works everywhere.
	Approved State = "approved"
	// Declined: denied, or the request expired (the hub drops unanswered
	// requests after a day). The token is dead.
	Declined State = "declined"
	// Unknown: no answer this time (network trouble); ask again.
	Unknown State = "unknown"
)

// Every is how often to ask (the docs say every 2–3 s).
const Every = 2500 * time.Millisecond

// Evaluate reads one /api/me answer. code is the four digits, while pending.
func Evaluate(me *hub.Me, err error) (s State, code string) {
	switch {
	case err != nil:
		return Unknown, ""
	case me == nil:
		return Unknown, ""
	case me.Device == nil:
		return Declined, ""
	case me.Device.Pending:
		return Pending, me.Device.Code
	}
	return Approved, ""
}

// Done reports whether s ends the wait.
func (s State) Done() bool { return s == Approved || s == Declined }

// Wait asks me every interval until the request is approved or declined,
// or ctx ends. update, if set, hears every state change (and a new code).
func Wait(ctx context.Context, me func(context.Context) (*hub.Me, error), every time.Duration,
	update func(s State, code string)) (State, error) {
	last, lastCode := State(""), ""
	for {
		actx, cancel := context.WithTimeout(ctx, 15*time.Second)
		m, err := me(actx)
		cancel()
		if ctx.Err() != nil {
			return last, ctx.Err()
		}
		s, code := Evaluate(m, err)
		if s == Unknown && last != "" {
			s = last // a hiccup doesn't change where the request stands
			code = lastCode
		}
		if (s != last || code != lastCode) && update != nil {
			update(s, code)
		}
		last, lastCode = s, code
		if s.Done() {
			return s, nil
		}
		t := time.NewTimer(every)
		select {
		case <-ctx.Done():
			t.Stop()
			return last, ctx.Err()
		case <-t.C:
		}
	}
}

// ErrDeclined is a join request that was denied or expired.
var ErrDeclined = errors.New("the request to join was declined (or expired)")

File: agent/droplet_agent/pairing.py
"""Waiting to be let in: a new device on the LAN asks, and one of yours answers.

`POST /api/device` from the LAN comes back pending, with a four-digit code
that the hub also shows on your other devices. Then `GET /api/me` (with the
new token) says how it went (docs/local-first.md §4):

- `device.pending` set: still waiting;
- `device` without `pending`: allowed in;
- `device: null`: denied, or the request expired.
"""

from __future__ import annotations

import time

from .hub import HubError, PinMismatch

PENDING = "pending"
APPROVED = "approved"
DENIED = "denied"
TIMED_OUT = "timed out"

POLL_EVERY = 2.5
MAX_ERRORS = 12       # in a row, about half a minute: the hub is gone, not just slow
WAIT_AT_MOST = 15 * 60  # the hub keeps the request for a day; setup can pick it up again


def state_of(device: dict | None) -> str:
    """What /api/me's `device` means for a device waiting to be let in."""
    if not isinstance(device, dict):
        return DENIED
    return PENDING if device.get("pending") else APPROVED


def wait_for_approval(check, *, every: float = POLL_EVERY, max_errors: int = MAX_ERRORS,
                      timeout: float = WAIT_AT_MOST, sleep=time.sleep, clock=time.monotonic,
                      on_error=None) -> str:
    """Poll `check()` (→ /api/me's device) until approved, denied or `timeout` seconds pass.

    A hub that can't be reached for a moment doesn't end the wait;
    `max_errors` failures in a row do (the last error is raised). A
    certificate that isn't the pinned one ends it at once.
    """
    end = clock() + timeout
    errors = 0
    while True:
        try:
            state = state_of(check())
            errors = 0
        except PinMismatch:
            raise
        except HubError as e:
            errors += 1
            if errors >= max_errors:
                raise
            if on_error is not None:
                on_error(e)
            state = PENDING
        if state != PENDING:
            return state
        if clock() >= end:
            return TIMED_OUT
        sleep(every)
